Sources/SwiftFormatPrettyPrint/TokenStreamCreator.swift

```diff
@@ -1794,26 +1862,43 @@ private final class TokenStreamCreator: SyntaxVisitor {
         tokens[tokens.count - 1] = .newlines(count, discretionary: true)
         return
 
-      // If we see two neighboring pairs of required newlines, combine them into a new token with
-      // the sum of their counts.
+      // If we see a pair of required newlines, combine them into a new token with the sum of
+      // their counts.
       case (.newlines(let first, discretionary: true), .newlines(let second, discretionary: true)):
         tokens[tokens.count - 1] = .newlines(first + second, discretionary: true)
         return
 
-      // If we see two neighboring pairs of non-required newlines, keep only the larger one.
+      // If we see a pair of non-required newlines, keep only the larger one.
       case (
         .newlines(let first, discretionary: false),
         .newlines(let second, discretionary: false)
       ):
         tokens[tokens.count - 1] = .newlines(max(first, second), discretionary: true)
         return
 
+      // If we see a pair of spaces where one or both are flexible, combine them into a new token
+      // with the maximum of their counts.
+      case (.space(let first, let firstFlexible), .space(let second, let secondFlexible))
```

Comment:
Is this the only place where the flexible value affects behavior? I'm not exactly clear on why `flexible` is needed.

Reply:
The problem I ran into was if you have something like this:

```swift
let a = b + c +  // foo
  d + e
```

and the formatting token after the last `+` on the first line is a space, rather than a break. Normally we *wouldn't* allow the line to break there at all if there was a space between the `+` and the `d`, but the trailing comment interferes with that. Rather than try to do something clever and shuffle the tokens or the comment around, we just allow it. But, that means there are two space tokens in the stream: `.space(1)` (after the `+`) and `.space(2)` (preceding the comment). There are various ways we could have handled this, but what I thought was the cleanest was to make the two spaces preceding the comment "flexible" in the sense that it will absorb neighboring spaces if they're present to ensure that the comment is preceded by *at most* two.

Another approach would have been to push the decision off until the pretty printer, but that seemed less clean; it's nicer to have the token stream creator coalesce neighboring tokens so that all the pretty printer does is print what it gets.